server/src/main/java/org/activityinfo/server/endpoint/rest/RootResource.java

```diff
@@ -130,8 +152,39 @@ public LocationsResource getLocations() {
     }
 
     @Path("/form/{id}")
-    public FormResource getForm(@PathParam("id") String id) {
-        return new FormResource(id);
+    public FormResource getForm(@PathParam("id") ResourceId id) {
+        return new FormResource(id, queryExecutor, userProvider);
+    }
+    
+    @Path("/query")
+    public QueryResource query() {
+        return new QueryResource(queryExecutor);
+    }
+    
+    @POST
+    @Path("/update") 
+    @Consumes(MediaType.APPLICATION_JSON)
+    public Response update(String json) {
+        
+        if(!DeploymentEnvironment.isAppEngineDevelopment() && !user.get().getEmail().endsWith("@bedatadriven.com")) {
+            return Response.status(Status.FORBIDDEN).build();
+        }
+        
+        Gson gson = new Gson();
+        final JsonElement jsonElement = gson.fromJson(json, JsonElement.class);
+
+        return queryExecutor.doWork(new HibernateQueryExecutor.StoreSession<Response>() {
+            @Override
+            public Response execute(CollectionCatalog catalog) {
+                Updater updater = new Updater(catalog);
+                try {
+                    updater.execute(jsonElement.getAsJsonObject());
+                } catch (InvalidUpdateException e) {
+                    throw new WebApplicationException(Response.status(Status.BAD_REQUEST).entity(e.getMessage()).build());
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Either log or rethrow this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1166)